core/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/OriginalListChangeIterator.java

```diff
@@ -77,4 +87,28 @@ static <Solution_> Move<Solution_> buildChangeMove(ListVariableStateSupply<Solut
         }
     }
 
+    /**
+     * This method looks for an element where the value matches the entity value range.
+     * Filtering is necessary when using the entity value range for list variables.
+     *
+     * @param upcomingLeftValue never null
+     * @param destinationIterator never null
+     * @param listVariableDescriptor never null
+     * @return null if no valid destination was found, at which point the iterator is exhausted.
+     */
+    private static ElementPosition findUnpinnedAndValidDestination(Object upcomingLeftValue,
+            Iterator<ElementPosition> destinationIterator, ListVariableDescriptor listVariableDescriptor) {
+        while (destinationIterator.hasNext()) {
+            var destination = destinationIterator.next();
+            if (isUnpinned(destination, listVariableDescriptor) && destination instanceof PositionInList destinationElement) {
+                var valueRange = listVariableDescriptor.getValueRangeDescriptor().extractValueRange(null,
```

Comment:
Why do we keep extracting these ranges? They should already be stored somewhere. They _never_ change.